Add Regexp feature

ChameleonTartu opened the pull request.

ChameleonTartu commented:
- [ ] Merge with the latest changes from `master`
- [ ] Finalize conceptual and implementation part for regexp feature
- [ ] Add more tests
- [ ] Fix coverage
- [ ] Add docs and examples

thomasborgen commented on piri/constants.py:
```diff
@@ -20,6 +20,12 @@
+REGEXP: Final[str] = 'regexp'
+SEARCH: Final[str] = 'search'
+GROUP: Final[str] = 'group'
+DEFAULT_GROUP: Final[int] = 0
```
default here is a great idea, lets do more of these for all the other ones.

ChameleonTartu commented:
@thomasborgen I propose to keep Regexp as simple as possible so that it will be:

```
{
  "search": "Example string",
  "group": [0, 1, 2, 3]
}
```

`"group": []` will correspond to all elements.

Therefore, there would be 3 main scenarios:
- Get a particular element by index
- Get a list of elements by indexes (in any order)
- Get all elements

To enable features such as slicing on the array, I propose to loosen slicing to accept arrays and strings (not only strings).

Please, let me know your thoughts on this.

thomasborgen commented:
That sounds really reasonable.
Default value will then be `[0]` right?

With the introduction of find `any value` feature, an issue should be created to also handle arrays in the slicing function.

ChameleonTartu commented:
> That sounds really reasonable.
> Default value will then be `[0]`, right?
> 
> With the introduction of the find `any value` feature, an issue should be created to handle arrays in the slicing function.

I thought the default to be `0`, not `[0]`. If we have Regexp applied to all values, they will become arrays independently if Regexp was called or not.

I thought about two main scenarios:
- Single element if Regexp couldn't find a match, not applied or `group` is an integer
- `group` is an array, and we return all elements as arrays, independently if there are no elements, single-element array, or all values

thomasborgen commented:
Yes of course you are right. 
Default value should be the one that gives back the first match.

thomasborgen commented on piri/schema.json:
```diff
@@ -95,6 +109,10 @@
                     "$ref": "#definitions/slicing",
                     "default": {}
                 },
+                "regexp": {
+                    "$ref": "#definitions/regexp",
+                    "default": {}
+                },
```
please move it above "slicing"

thomasborgen commented on tests/functions/test_apply_regexp.py:
```diff
@@ -0,0 +1,17 @@
+from piri.functions import apply_regexp
+
+
+def test_regexp_on_index_out_of_range():
+    """Test regexp when group indeces are out of range."""
+    with pytest.raises(IndexError):
+        assert apply_regexp(
+            'Hard work',
+            {'search': 'r', 'group': [1, 2, 3]},
+        )
```
If our apply_regexp function can raise exceptions it should be @safe annotated. The result here should be a Failure instance of IndexError.   == Failure(IndexError('out of range'))

Please add this test to test_handle_mapping.py as well. We must know how it will work with the rest of handle_attribute flow code. Add if statements with a targe = None, check so that we make sure that changing the value will work even when regexp fails or raises exception

thomasborgen commented on piri/functions.py:
```diff
@@ -176,6 +181,76 @@ def apply_slicing(
     return value_to_slice[slicing[FROM]:slicing.get(TO)]
 
 
+def apply_regexp(  # noqa: WPS212, WPS234
```
If this function can throw exception, it must be @safe annotated

thomasborgen commented on piri/handlers.py:
```diff
@@ -61,6 +63,9 @@ def handle_mapping(
         partial(fetch_data_by_keys, path=cfg.get(PATH, [])),
         fix(lambda _: None),  # type: ignore
+        map_(partial(
+            apply_regexp, regexp=cfg.get(REGEXP, []),
```
if the apply_regexp fails and returns a Failure() then we must "fix" it like the fix above. This also means that in the apply_regexp function you can remove "if matches:" check since its completely fine to get an exception that we will just handle with fix()

ChameleonTartu commented:
@thomasborgen, I think I addressed all the comments in the PR. Can you take a look at it? I will need some feedback from you before continuing/merging this PR.

thomasborgen commented on piri/schema.json:
```diff
@@ -51,6 +51,20 @@
+            "type": "object",
+            "properties": {
+                "search": {
+                    "type": "string"
```
@ChameleonTartu did this not work?

thomasborgen reviewed (approved):
Looks really good. Thanks a lot for all the work you've put into this. This will be a really important feature for the library
